### PR TITLE
examples/dmabuf-capture: move encoding to a separate thread

### PR DESCRIPTION
**Pull request by atomnuker**
Drop new frames if too slow. Speeds up encoding significantly, even with vaapi.

### CODE REVIEW

File: examples/meson.build [hunk 1]
```diff
@@ -47,6 +47,7 @@ if libavutil.found() and libavcodec.found() and libavformat.found()
 	executable(
 		'dmabuf-capture',
 		'dmabuf-capture.c',
-		dependencies: [wayland_client, wlr_protos, libavutil, libavcodec, libavformat, wlroots]
+		dependencies: [wayland_client, wlr_protos, libavutil, libavcodec, libavformat, wlroots],
+		link_args : '-pthread',
```
Comment (ascent12) [inline examples/meson.build]: Meson has a proper way you're supposed to link against pthreads: http://mesonbuild.com/Threads.html

File: examples/dmabuf-capture.c [hunk 2]
```diff
@@ -66,6 +79,75 @@ struct capture_context {
 	float out_bitrate;
 };
 
+static int init_fifo(FifoBuffer *buf, int max_queued_frames)
+{
```
Comment (SirCmpwn) [inline examples/dmabuf-capture.c]: style, move this to the previous line

File: examples/dmabuf-capture.c [hunk 3]
```diff
@@ -24,6 +26,15 @@ struct wayland_output {
...
 };
 
+typedef struct FifoBuffer {
+	AVFrame **queued_frames;
+	int num_queued_frames;
+	int max_queued_frames;
+	pthread_mutex_t lock;
+	pthread_cond_t cond;
+	pthread_mutex_t cond_lock;
+} FifoBuffer;
```
Comment (SirCmpwn) [inline examples/dmabuf-capture.c]: Don't use typedefs

Comment (SirCmpwn) [reply examples/dmabuf-capture.c]: e.g. just do `struct fifo_buffer`

File: examples/dmabuf-capture.c [hunk 4]
```diff
@@ -66,6 +79,75 @@ struct capture_context {
...
 };
 
+static int init_fifo(FifoBuffer *buf, int max_queued_frames)
+{
+	pthread_mutex_init(&buf->lock, NULL);
+	pthread_cond_init(&buf->cond, NULL);
+	pthread_mutex_init(&buf->cond_lock, NULL);
+	buf->num_queued_frames = 0;
+	buf->max_queued_frames = max_queued_frames;
+	buf->queued_frames = av_mallocz(buf->max_queued_frames*sizeof(AVFrame));
```
Comment (SirCmpwn) [inline examples/dmabuf-capture.c]: spaces around operators

File: examples/dmabuf-capture.c [hunk 5]
```diff
@@ -66,6 +79,75 @@ struct capture_context {
...
+
+	pthread_mutex_unlock(&buf->lock);
+	return rf;
+}
+
+static void free_fifo(FifoBuffer *buf)
+{
+	pthread_mutex_lock(&buf->lock);
+	if (buf->num_queued_frames)
+		for (int i = 0; i < buf->num_queued_frames; i++)
```
Comment (SirCmpwn) [inline examples/dmabuf-capture.c]: braces are mandatory

File: examples/meson.build [hunk 6]
```diff
@@ -4,6 +4,7 @@ wayland_cursor = dependency('wayland-cursor')
 libavutil = dependency('libavutil', required: false)
 libavcodec = dependency('libavcodec', required: false)
 libavformat = dependency('libavformat', required: false)
+threads = dependency('threads')
```
Comment (SirCmpwn) [inline examples/meson.build]: This is already done at the top of this file

**Comment by SirCmpwn** [comment]
LGTM aside from style issues

**Comment by atomnuker** [comment]
Should be fine now?

File: examples/dmabuf-capture.c [hunk 7]
```diff
@@ -24,6 +26,15 @@ struct wayland_output {
...
 };
 
+struct fifo_buffer {
+	AVFrame **queued_frames;
+	int num_queued_frames;
+	int max_queued_frames;
+	pthread_mutex_t lock;
+	pthread_cond_t cond;
+	pthread_mutex_t cond_lock;
+} fifo_buffer;
```
Comment (SirCmpwn) [inline examples/dmabuf-capture.c]: This creates a struct type named fifo_buffer and also a static instance of that struct called fifo_buffer

File: examples/dmabuf-capture.c [hunk 8]
```diff
@@ -332,33 +422,66 @@ static void frame_ready(void *data, struct zwlr_export_dmabuf_frame_v1 *frame,
...
-		enc_input = soft_frame;
+	if (push_to_fifo(&ctx->vid_frames, mapped_frame)) {
+		av_log(ctx, AV_LOG_WARNING, "Dropped frame!\n");
 	}
 
-	/* Nanoseconds */
-	enc_input->pts = (((uint64_t)tv_sec_hi) << 32) | tv_sec_lo;
-	enc_input->pts *= 1000000000;
-	enc_input->pts += tv_nsec;
+	if (!ctx->quit && !ctx->err)
```
Comment (emersion) [inline examples/dmabuf-capture.c]: Style: missing brackets here

File: examples/dmabuf-capture.c [hunk 9]
```diff
@@ -332,33 +422,66 @@ static void frame_ready(void *data, struct zwlr_export_dmabuf_frame_v1 *frame,
...
+}
+
+void *vid_encode_thread(void *arg) {
+	int err = 0;
+	struct capture_context *ctx = arg;
 
 	do {
-		err = avcodec_send_frame(ctx->avctx, enc_input);
+		AVFrame *f = NULL;
+		if (get_fifo_size(&ctx->vid_frames) || !ctx->quit)
```
Comment (emersion) [inline examples/dmabuf-capture.c]: Ditto

**Comment by atomnuker** [comment]
Nazis.

**Comment by SirCmpwn** [comment]
Thanks!

**Pull request merged**